src/v/redpanda/application.cc

```diff
@@ -1822,7 +1823,61 @@ void application::wire_up_and_start(::stop_signal& app_signal, bool test_mode) {
     const auto& node_uuid = storage.local().node_uuid();
     cluster::cluster_discovery cd(
       node_uuid, storage.local(), app_signal.abort_source());
-    auto node_id = cd.determine_node_id().get();
+
+    bool ever_ran_controller = storage.local()
+                                 .kvs()
+                                 .get(
+                                   storage::kvstore::key_space::controller,
+                                   bytes("configuration_invariants"))
+                                 .has_value();
+
+    model::node_id node_id;
+    if (config::node().node_id().has_value() && ever_ran_controller) {
+        vlog(
+          _log.debug,
+          "Running with already-established node ID {}",
+          config::node().node_id());
+        node_id = config::node().node_id().value();
+    } else {
+        auto registration_result = cd.determine_node_id().get();
+        node_id = registration_result.assigned_node_id;
+
+        if (registration_result.newly_registered) {
+            vlog(
+              _log.info,
+              "Registered with cluster as node ID {}",
+              registration_result.assigned_node_id);
+            if (registration_result.controller_snapshot.has_value()) {
+                // Do something with the controller snapshot
+                auto snap = serde::from_iobuf<cluster::controller_snapshot>(
```

Comment:
This looks dangerous to me - AFAICT, sync read/write are autogenerated and I am not sure that they are correct/compatible with async versions. We should probably somehow forbid usage of sync versions.

Reply:
I've changed this to use a specialized `controller_join_snapshot` structure that only includes the simpler sync-encoded structures (bootstrap, config,features,metrics reporter)

Would still be nice to make serde a bit safer for this case so that we're not relying on knowledge of  which structs have sync vs async encoding defined.